text/0000-autoref-copy.md

```diff
@@ -0,0 +1,199 @@
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    println!("{:?}", map.get(1)); // ERROR: expected `&{integer}`
+
+    // Instead, users have to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+This is an unnecessary frustration. This RFC proposes to prevent this issue
+by auto-referencing `Copy` types.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When calling a function which expects a reference to a `Copy` type, you can
+pass an owned `Copy` type and a reference will be created automatically:
+
+```rust
+fn print_num(x: &usize) { println!("{:?}", x); }
+
+fn main() {
+  let x = 5;
+  print_num(x);
+}
+```
+
+This is particularly convenient when working with some generic types whose
+methods take references as arguments:
+
+```rust
+use std::collections::HashMap;
+fn main() {
+    let mut map = HashMap::new();
+    map.insert(1, "hello!");
+
+    // We can write this:
+    println!("{:?}", map.get(1));
+
+    // Instead of having to write this:
+    println!("{:?}", map.get(&1));
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`Copy` types autoreference: at
+[coercion sites](https://github.com/rust-lang/rfcs/blob/master/text/0401-coercions.md#coercions),
+when a reference is expected, but an owned
+`Copy` type is found, an `&` will be automatically inserted.
+
+Some types have interior mutability through `UnsafeCell`, so passing them
+via a hidden reference could allow surprising mutations to occur.
+To prevent this, the autoreferencing coercion is limited to types which do not
+directly contain an internal `UnsafeCell`.
+Types which reference an `UnsafeCell` through an indirection can still be
+coerced (e.g. `&MyCopyCell<u8>` to `&&MyCopyCell<u8>`).
+
+This coercion will not occur if the lifetime of the resulting reference would
+last longer than the call site. For example:
+
+```rust
+// `foo`'s argument can be coerced to from `u8` because it only lasts for the
+// lifetime of the function call:
+fn foo(x: &u8) { ... }
+
+// `bar`'s argument cannot be coerced to from `u8` because the lifetime of the
+// required reference outlives the scope of `bar`.
+fn bar<'a>(x: &'a u8) -> &'a u8 { ... }
+
+fn main() {
+    foo(1); // OK
+
+    let x = bar(5); // ERROR: expected `&u8`, found `u8`.
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This increases the special behavior of `Copy` types, making it potentially
+confusing to new users. Some simple, non-generic code becomes more confusing
+because `fn foo(x: &i32) { ... }` can now be called like `foo(5)`;
+- The existing coercion mechanism doesn't allow for coercion of generic
+arguments, nor does it use coercions in trait lookup. Because of this, users
+will still need to manually add `&` when trying to, for example, index into
+a `HashMap<usize, T>`
+(users will still need to write `x[&5]` instead of `x[5]`).
```

Comment:
I think we *do* coercions in indexing? At least in the "single impl" case? You *can* write this today:

```Rust
use std::collections::HashMap;

fn main() {
    let s : HashMap<u32, u32> = HashMap::new();
    let i: &&u32 = &&0;
    s[i];
}
```

Reply:
You're absolutely right. I don't know why I thought we didn't. In particular, `AsRef` examples like `&Rc<i32>` work just fine:

```rust
use std::collections::HashMap;
use std::rc::Rc;

fn main() {
    let mut x = HashMap::new();
    x.insert(5, 5);
    x[&Rc::new(5)];
}
```

I will remove indexing from the examples. Thanks for catching me on that!